File: libraries/python/getpatter/providers/sarvam_tts.py
"""Sarvam AI TTS provider — HTTP REST one-shot endpoint, pure aiohttp.

Sarvam's Bulbul models synthesize 11 Indian languages (Hindi, Bengali, Tamil,
Telugu, Kannada, Malayalam, Marathi, Gujarati, Punjabi, Odia, Indian English)
plus code-mixed text. This adapter targets the REST one-shot endpoint
``POST https://api.sarvam.ai/text-to-speech`` which returns JSON of the form
``{"request_id": "...", "audios": ["<base64-audio>", ...]}`` — we base64-decode
each entry and yield the raw bytes, mapping cleanly onto Patter's
``TTSProvider.synthesize(text) -> AsyncIterator[bytes]`` contract with no vendor
SDK.

Sarvam also exposes a WebSocket streaming transport
(``wss://api.sarvam.ai/text-to-speech/ws?model=<model_id>``) with sub-250 ms
first-byte; like the Cartesia adapter we use the REST path because it already
meets Patter's TTFB target while keeping the dependency surface minimal (just
``aiohttp``, no websocket client).

Auth is an API subscription key sent as the ``api-subscription-key`` HTTP
header (no Bearer scheme). The key is read from the ``api_key`` argument or the
``SARVAM_API_KEY`` environment variable.

Default config requests ``codec="linear16"`` (raw PCM_S16LE) at 16 kHz so the
output drops straight into the Patter pipeline without transcoding — the stream
handler does the final 16 kHz -> 8 kHz mu-law step for the carrier. For real
phone calls the :meth:`for_twilio` / :meth:`for_telnyx` factories request
``mulaw`` @ 8 kHz directly (Sarvam supports both natively), so the pipeline
takes the carrier-native passthrough path with zero resampling.

Bulbul v3 is the default model (latest; 35+ voices; ``temperature`` control).
Pass ``model="bulbul:v2"`` for the legacy generation, which instead exposes
``pitch`` / ``loudness`` / ``enable_preprocessing``. Those controls are gated
per-model so we never send fields the selected model rejects.
"""

from __future__ import annotations

import base64
import logging
import os
from enum import IntEnum, StrEnum
from typing import ClassVar, Any, AsyncIterator, Optional, Union
from urllib.parse import urlsplit

from getpatter.providers.base import TTSProvider

logger = logging.getLogger("getpatter.providers.sarvam_tts")

try:  # pragma: no cover - trivial import guard
    import aiohttp
except ImportError:  # pragma: no cover
    aiohttp = None  # type: ignore

SARVAM_BASE_URL = "https://api.sarvam.ai/text-to-speech"


class SarvamModel(StrEnum):
    """Sarvam Bulbul TTS model families."""

    BULBUL_V3 = "bulbul:v3"
    BULBUL_V2 = "bulbul:v2"


class SarvamAudioCodec(StrEnum):
    """``output_audio_codec`` values accepted by the REST API.

    ``LINEAR16`` is headerless 16-bit PCM (the pipeline-friendly default);
    ``MULAW`` / ``ALAW`` are G.711 telephony codecs.
    """

    WAV = "wav"
    MP3 = "mp3"
    LINEAR16 = "linear16"
    MULAW = "mulaw"
    ALAW = "alaw"
    OPUS = "opus"
    FLAC = "flac"
    AAC = "aac"


class SarvamLanguage(StrEnum):
    """BCP-47 ``target_language_code`` values for the 11 supported languages.

    Note Sarvam uses ``od-IN`` for Odia (not the more common ``or-IN``).
    """

    HINDI = "hi-IN"
    BENGALI = "bn-IN"
    TAMIL = "ta-IN"
    TELUGU = "te-IN"
    KANNADA = "kn-IN"
    MALAYALAM = "ml-IN"
    MARATHI = "mr-IN"
    GUJARATI = "gu-IN"
    PUNJABI = "pa-IN"
    ODIA = "od-IN"
    ENGLISH = "en-IN"


class SarvamSampleRate(IntEnum):
    """Output sample rates (Hz) accepted by the REST ``speech_sample_rate``."""

    HZ_8000 = 8000
    HZ_16000 = 16000
    HZ_22050 = 22050
    HZ_24000 = 24000
    HZ_32000 = 32000
    HZ_44100 = 44100
    HZ_48000 = 48000


class SarvamTTS(TTSProvider):
    """Sarvam AI TTS over the REST ``/text-to-speech`` one-shot endpoint.

    Output is base64 audio decoded to raw bytes. With the default
    ``codec="linear16"`` at 16 kHz these are PCM_S16LE samples that line up
    with Patter's pipeline without a resample step.

    Telephony optimization
    ----------------------
    The constructor default ``sample_rate=16000`` is correct for web playback
    and 16 kHz pipelines. For real phone calls use the carrier-specific
    factories instead:

    * :meth:`for_twilio` / :meth:`for_telnyx` — emit ``mulaw`` @ 8 kHz, the
      carrier wire codec, so the pipeline skips resampling AND PCM -> mu-law
      encoding (bit-clean passthrough). Sarvam supports ``mulaw`` + 8 kHz
      natively, so no transcoding is needed on either side.
    """

    #: Stable pricing/dashboard key — read by stream-handler/metrics.
    provider_key: ClassVar[str] = "sarvam"

    def __init__(
        self,
        api_key: Optional[str] = None,
        *,
        model: Union[SarvamModel, str] = SarvamModel.BULBUL_V3,
        speaker: str = "shubh",
        language: Union[SarvamLanguage, str] = SarvamLanguage.ENGLISH,
        codec: Union[SarvamAudioCodec, str] = SarvamAudioCodec.LINEAR16,
        sample_rate: Union[SarvamSampleRate, int] = SarvamSampleRate.HZ_16000,
        pace: Optional[float] = None,
        pitch: Optional[float] = None,
        loudness: Optional[float] = None,
        temperature: Optional[float] = None,
        enable_preprocessing: Optional[bool] = None,
        dict_id: Optional[str] = None,
        base_url: str = SARVAM_BASE_URL,
        session: Optional["aiohttp.ClientSession"] = None,
    ) -> None:
        if aiohttp is None:
            raise ImportError(
                "aiohttp is required for SarvamTTS. "
                "Install with: pip install getpatter[sarvam]"
            )

        resolved_key = api_key or os.environ.get("SARVAM_API_KEY")
        if not resolved_key:
            raise ValueError(
                "Sarvam API key is required, either as argument or set "
                "SARVAM_API_KEY environment variable"
            )

        self.api_key = resolved_key
        self.model = model
        self.speaker = speaker
        self.language = language
        self.codec = codec
        self.sample_rate = sample_rate
        self.pace = pace
        self.pitch = pitch
        self.loudness = loudness
        self.temperature = temperature
        self.enable_preprocessing = enable_preprocessing
        self.dict_id = dict_id
        self.base_url = base_url
        self._owns_session = session is None
        self._session = session

    def __repr__(self) -> str:
        # Never leak the API key in repr / logs.
        return (
            f"SarvamTTS(model={self.model!r}, speaker={self.speaker!r}, "
            f"language={self.language!r}, codec={self.codec!r}, "
            f"sample_rate={self.sample_rate})"
        )

    # ------------------------------------------------------------------
    # Telephony factories
    # ------------------------------------------------------------------

    @classmethod
    def for_twilio(
        cls,
        api_key: Optional[str] = None,
        **kwargs: Any,
    ) -> "SarvamTTS":
        """Build an instance pre-configured for Twilio Media Streams.

        Emits ``mulaw`` @ 8 kHz — exactly Twilio's wire codec — so the
        pipeline takes the passthrough path: zero resampling, zero PCM ->
        mu-law encoding. Sarvam supports mu-law / 8 kHz natively, so this is a
        true carrier-native path (no resample on Sarvam's side either).
        """
        kwargs.pop("sample_rate", None)
        kwargs.pop("codec", None)
        return cls(
            api_key=api_key,
            codec=SarvamAudioCodec.MULAW,
            sample_rate=SarvamSampleRate.HZ_8000,
            **kwargs,
        )

    @classmethod
    def for_telnyx(
        cls,
        api_key: Optional[str] = None,
        **kwargs: Any,
    ) -> "SarvamTTS":
        """Build an instance pre-configured for Telnyx bidirectional media.

        Telnyx pins the wire to PCMU/mu-law @ 8 kHz, so emitting ``mulaw`` @
        8 kHz flows end-to-end with zero resampling — the same passthrough win
        as :meth:`for_twilio`.
        """
        kwargs.pop("sample_rate", None)
        kwargs.pop("codec", None)
        return cls(
            api_key=api_key,
            codec=SarvamAudioCodec.MULAW,
            sample_rate=SarvamSampleRate.HZ_8000,
            **kwargs,
        )

    def source_audio_format(self) -> "AudioFormat":
        """Declare the audio format this adapter emits so the pipeline sender
        derives the correct resample ratio (or skips it for mu-law / a-law
        passthrough) instead of assuming a fixed 16 kHz source. See
        ``getpatter.audio.format``.
        """
        from getpatter.audio.format import AudioFormat

        codec = str(self.codec)
        if codec == SarvamAudioCodec.MULAW.value:
            return AudioFormat(encoding="mulaw", sample_rate=int(self.sample_rate))
        if codec == SarvamAudioCodec.ALAW.value:
            return AudioFormat(encoding="alaw", sample_rate=int(self.sample_rate))
        return AudioFormat(encoding="pcm_s16le", sample_rate=int(self.sample_rate))

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _is_v2(self) -> bool:
        """True when the configured model is the legacy ``bulbul:v2``."""
        return str(self.model) == SarvamModel.BULBUL_V2.value

    def _ensure_session(self) -> "aiohttp.ClientSession":
        if self._session is None:
            self._session = aiohttp.ClientSession()
            self._owns_session = True
        return self._session

    def _build_payload(self, text: str) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "text": text,
            "target_language_code": str(self.language),
            "model": str(self.model),
            "speaker": self.speaker,
            "output_audio_codec": str(self.codec),
            "speech_sample_rate": int(self.sample_rate),
        }

        if self.pace is not None:
            payload["pace"] = self.pace

        # Per-model controls: pitch / loudness / enable_preprocessing are
        # bulbul:v2-only; temperature / dict_id are bulbul:v3-only. Gate them so
        # we never POST a field the selected model rejects.
        if self._is_v2():
            if self.pitch is not None:
                payload["pitch"] = self.pitch
            if self.loudness is not None:
                payload["loudness"] = self.loudness
            if self.enable_preprocessing is not None:
                payload["enable_preprocessing"] = self.enable_preprocessing
        else:
            if self.temperature is not None:
                payload["temperature"] = self.temperature
            if self.dict_id is not None:
                payload["dict_id"] = self.dict_id

        return payload

    def _record_synthesis_cost(self, text: str) -> None:
        """Emit ``patter.cost.tts_chars`` for the synthesised text.

        NO PII: only the character COUNT is recorded, never the text itself.
        """
        try:
            from getpatter.observability.attributes import record_patter_attrs

            record_patter_attrs(
                {
                    "patter.cost.tts_chars": len(text),
                    "patter.tts.provider": "sarvam",
                }
            )
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("_record_synthesis_cost failed", exc_info=True)

    async def synthesize(self, text: str) -> AsyncIterator[bytes]:
        """Synthesize ``text`` and yield decoded audio bytes.

        With the default ``codec="linear16"`` these are raw PCM_S16LE samples
        at ``sample_rate``. The REST endpoint returns the whole clip in one
        JSON response (``audios`` is a list of base64 strings); we decode and
        yield each entry in order.
        """
        self._record_synthesis_cost(text)
        session = self._ensure_session()

        headers = {
            "Content-Type": "application/json",
            "api-subscription-key": self.api_key,
        }

        async with session.post(
            self.base_url,
            headers=headers,
            json=self._build_payload(text),
            timeout=aiohttp.ClientTimeout(total=30),
        ) as resp:
            if resp.status != 200:
                body = await resp.text()
                raise RuntimeError(f"Sarvam TTS error {resp.status}: {body[:500]}")
            data = await resp.json()
            for audio_b64 in _iter_audio_b64(data):
                try:
                    decoded = base64.b64decode(audio_b64)
                except (ValueError, TypeError):
                    continue
                if decoded:
                    yield decoded

    async def warmup(self) -> None:
        """Pre-call HTTP warmup for the Sarvam TTS API.

        Issues a lightweight ``GET`` against the API origin so DNS + TLS +
        HTTP/2 are already up by the time the first :meth:`synthesize` POST
        lands. Best-effort: 5 s timeout, all exceptions swallowed at DEBUG.

        Billing safety: Sarvam does not document a free voice/metadata GET, so
        this only primes the connection against the API origin root — it never
        hits the synthesis endpoint, so no characters are billed. The actual
        synthesis is billed only when ``POST /text-to-speech`` runs with
        non-empty ``text``.

        Note: like the Cartesia / Inworld adapters this uses the HTTP path
        (Sarvam also exposes a WebSocket transport we don't use here), so the
        latency win is the HTTP connection prime (~50-150 ms) rather than a WS
        pre-handshake.
        """
        try:
            session = self._ensure_session()
            parts = urlsplit(self.base_url)
            origin = f"{parts.scheme}://{parts.netloc}/"
            headers = {"api-subscription-key": self.api_key}
            async with session.get(
                origin,
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=5),
            ) as resp:
                # Drain so the underlying connection returns cleanly to the pool.
                await resp.read()
        except Exception as exc:  # noqa: BLE001 - best-effort
            logger.debug("Sarvam TTS warmup failed (best-effort): %s", exc)

    async def close(self) -> None:
        """Close the underlying session (idempotent)."""
        if self._session is not None and self._owns_session:
            await self._session.close()
            self._session = None


def _iter_audio_b64(data: Any) -> list[str]:
    """Extract the base64 audio strings from a Sarvam REST response.

    Accepts the documented ``{"audios": [...]}`` shape and is defensive about a
    single-string ``audios`` value. Returns ``[]`` for anything unexpected so
    the caller yields nothing rather than raising on a malformed body.
    """
    if not isinstance(data, dict):
        return []
    audios = data.get("audios")
    if isinstance(audios, str):
        return [audios] if audios else []
    if isinstance(audios, list):
        return [a for a in audios if isinstance(a, str) and a]
    return []

File: libraries/typescript/src/providers/sarvam-tts.ts
/**
 * Sarvam AI TTS provider — HTTP REST one-shot endpoint.
 *
 * Sarvam's Bulbul models synthesize 11 Indian languages (Hindi, Bengali,
 * Tamil, Telugu, Kannada, Malayalam, Marathi, Gujarati, Punjabi, Odia, Indian
 * English) plus code-mixed text. This adapter targets the REST one-shot
 * endpoint `POST https://api.sarvam.ai/text-to-speech`, which returns JSON of
 * the form `{ "request_id": "...", "audios": ["<base64-audio>", ...] }` — we
 * base64-decode each entry and yield the raw bytes, mapping cleanly onto
 * Patter's `synthesize(text)` contract with no vendor SDK (just `fetch`).
 *
 * Sarvam also exposes a WebSocket streaming transport
 * (`wss://api.sarvam.ai/text-to-speech/ws?model=<model_id>`); like the Cartesia
 * adapter we use the REST path because it already meets Patter's TTFB target
 * while keeping the provider dependency-free.
 *
 * Auth is an API subscription key sent as the `api-subscription-key` HTTP
 * header (no Bearer scheme), read from the constructor or `SARVAM_API_KEY`.
 *
 * The default config requests `codec="linear16"` (raw PCM_S16LE) at 16 kHz so
 * the output drops straight into the Patter pipeline without transcoding. For
 * real phone calls the {@link SarvamTTS.forTwilio} / {@link SarvamTTS.forTelnyx}
 * factories request `mulaw` @ 8 kHz directly (Sarvam supports both natively),
 * so the pipeline takes the carrier-native passthrough path.
 *
 * Bulbul v3 is the default model (latest; 35+ voices; `temperature` control).
 * Pass `model: "bulbul:v2"` for the legacy generation, which instead exposes
 * `pitch` / `loudness` / `enablePreprocessing`. Those controls are gated
 * per-model so we never send fields the selected model rejects.
 */

import { getLogger } from '../logger';

const SARVAM_BASE_URL = 'https://api.sarvam.ai/text-to-speech';

/** Sarvam Bulbul TTS model families. */
export const SarvamModel = {
  BULBUL_V3: 'bulbul:v3',
  BULBUL_V2: 'bulbul:v2',
} as const;
export type SarvamModel = (typeof SarvamModel)[keyof typeof SarvamModel];

/**
 * `output_audio_codec` values accepted by the REST API. `LINEAR16` is
 * headerless 16-bit PCM (the pipeline-friendly default); `MULAW` / `ALAW` are
 * G.711 telephony codecs.
 */
export const SarvamAudioCodec = {
  WAV: 'wav',
  MP3: 'mp3',
  LINEAR16: 'linear16',
  MULAW: 'mulaw',
  ALAW: 'alaw',
  OPUS: 'opus',
  FLAC: 'flac',
  AAC: 'aac',
} as const;
export type SarvamAudioCodec = (typeof SarvamAudioCodec)[keyof typeof SarvamAudioCodec];

/**
 * BCP-47 `target_language_code` values for the 11 supported languages. Note
 * Sarvam uses `od-IN` for Odia (not the more common `or-IN`).
 */
export const SarvamLanguage = {
  HINDI: 'hi-IN',
  BENGALI: 'bn-IN',
  TAMIL: 'ta-IN',
  TELUGU: 'te-IN',
  KANNADA: 'kn-IN',
  MALAYALAM: 'ml-IN',
  MARATHI: 'mr-IN',
  GUJARATI: 'gu-IN',
  PUNJABI: 'pa-IN',
  ODIA: 'od-IN',
  ENGLISH: 'en-IN',
} as const;
export type SarvamLanguage = (typeof SarvamLanguage)[keyof typeof SarvamLanguage];

/** Output sample rates (Hz) accepted by the REST `speech_sample_rate`. */
export const SarvamSampleRate = {
  HZ_8000: 8000,
  HZ_16000: 16000,
  HZ_22050: 22050,
  HZ_24000: 24000,
  HZ_32000: 32000,
  HZ_44100: 44100,
  HZ_48000: 48000,
} as const;
export type SarvamSampleRate = (typeof SarvamSampleRate)[keyof typeof SarvamSampleRate];

/** Constructor options for {@link SarvamTTS}. */
export interface SarvamTTSOptions {
  /** Model id. Defaults to `"bulbul:v3"`. */
  model?: SarvamModel | string;
  /** Speaker name (lowercase). Defaults to `"shubh"` (a bulbul:v3 voice). */
  speaker?: string;
  /** BCP-47 language code, e.g. `"hi-IN"`, `"ta-IN"`, `"en-IN"`. Default `"en-IN"`. */
  language?: SarvamLanguage | string;
  /**
   * Output audio codec. Default `linear16` (raw PCM16). Set `mulaw` with
   * `sampleRate=8000` to emit Twilio/Telnyx wire-native μ-law directly — the
   * pipeline then skips resampling and PCM → μ-law encoding entirely (see
   * {@link SarvamTTS.forTwilio}).
   */
  codec?: SarvamAudioCodec | string;
  /** Output sample rate in Hz. Defaults to 16000. */
  sampleRate?: SarvamSampleRate | number;
  /** Speaking pace multiplier (v3: 0.5–2.0, v2: 0.3–3.0; default 1.0 server-side). */
  pace?: number;
  /** Pitch shift (bulbul:v2 ONLY; −0.75 to 0.75). Ignored on v3. */
  pitch?: number;
  /** Loudness multiplier (bulbul:v2 ONLY; 0.3–3.0). Ignored on v3. */
  loudness?: number;
  /** Sampling temperature (bulbul:v3 ONLY; 0.01–2.0). Ignored on v2. */
  temperature?: number;
  /** Normalise numbers/entities (bulbul:v2 ONLY). Ignored on v3. */
  enablePreprocessing?: boolean;
  /** Custom pronunciation dictionary id (bulbul:v3 ONLY). Ignored on v2. */
  dictId?: string;
  /** Override the REST endpoint. */
  baseUrl?: string;
}

/** Audio format declared back to the pipeline sender via {@link SarvamTTS.sourceAudioFormat}. */
type SourceAudioFormat = { encoding: 'pcm_s16le' | 'mulaw' | 'alaw'; sampleRate: number };

/** Sarvam TTS provider backed by the HTTP REST `/text-to-speech` endpoint. */
export class SarvamTTS {
  /** Stable pricing/dashboard key — read by stream-handler/metrics. */
  static readonly providerKey = 'sarvam';
  private readonly apiKey: string;
  private readonly model: string;
  private readonly speaker: string;
  private readonly language: string;
  private readonly codec: string;
  private readonly sampleRate: number;
  private readonly pace?: number;
  private readonly pitch?: number;
  private readonly loudness?: number;
  private readonly temperature?: number;
  private readonly enablePreprocessing?: boolean;
  private readonly dictId?: string;
  private readonly baseUrl: string;

  constructor(apiKey: string, opts: SarvamTTSOptions = {}) {
    if (!apiKey) {
      throw new Error('Sarvam TTS: apiKey is required');
    }
    this.apiKey = apiKey;
    this.model = opts.model ?? SarvamModel.BULBUL_V3;
    this.speaker = opts.speaker ?? 'shubh';
    this.language = opts.language ?? SarvamLanguage.ENGLISH;
    this.codec = opts.codec ?? SarvamAudioCodec.LINEAR16;
    this.sampleRate = opts.sampleRate ?? SarvamSampleRate.HZ_16000;
    this.pace = opts.pace;
    this.pitch = opts.pitch;
    this.loudness = opts.loudness;
    this.temperature = opts.temperature;
    this.enablePreprocessing = opts.enablePreprocessing;
    this.dictId = opts.dictId;
    this.baseUrl = opts.baseUrl ?? SARVAM_BASE_URL;
  }

  /**
   * Declare the audio format this adapter emits, so the pipeline sender can
   * derive the correct resample ratio (or skip it for μ-law/a-law passthrough)
   * instead of assuming a fixed 16 kHz source. See `audio/format.ts`.
   */
  sourceAudioFormat(): SourceAudioFormat {
    if (this.codec === SarvamAudioCodec.MULAW) {
      return { encoding: 'mulaw', sampleRate: this.sampleRate };
    }
    if (this.codec === SarvamAudioCodec.ALAW) {
      return { encoding: 'alaw', sampleRate: this.sampleRate };
    }
    return { encoding: 'pcm_s16le', sampleRate: this.sampleRate };
  }

  /**
   * Construct an instance pre-configured for Twilio Media Streams.
   *
   * Emits `mulaw` @ 8 kHz — exactly Twilio's wire codec — so the pipeline takes
   * the passthrough path: zero resampling, zero PCM → μ-law encoding. Sarvam
   * supports μ-law / 8 kHz natively, so this is a true carrier-native path.
   */
  static forTwilio(
    apiKey: string,
    options: Omit<SarvamTTSOptions, 'codec' | 'sampleRate'> = {},
  ): SarvamTTS {
    return new SarvamTTS(apiKey, {
      ...options,
      codec: SarvamAudioCodec.MULAW,
      sampleRate: SarvamSampleRate.HZ_8000,
    });
  }

  /**
   * Construct an instance pre-configured for Telnyx bidirectional media.
   *
   * Telnyx pins the wire to PCMU/μ-law @ 8 kHz, so emitting `mulaw` @ 8 kHz
   * flows end-to-end with zero resampling — the same passthrough win as
   * {@link SarvamTTS.forTwilio}.
   */
  static forTelnyx(
    apiKey: string,
    options: Omit<SarvamTTSOptions, 'codec' | 'sampleRate'> = {},
  ): SarvamTTS {
    return new SarvamTTS(apiKey, {
      ...options,
      codec: SarvamAudioCodec.MULAW,
      sampleRate: SarvamSampleRate.HZ_8000,
    });
  }

  /** True when the configured model is the legacy `bulbul:v2`. */
  private isV2(): boolean {
    return this.model === SarvamModel.BULBUL_V2;
  }

  /** Build the JSON payload for the Sarvam `/text-to-speech` endpoint. */
  private buildPayload(text: string): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      text,
      target_language_code: this.language,
      model: this.model,
      speaker: this.speaker,
      output_audio_codec: this.codec,
      speech_sample_rate: this.sampleRate,
    };

    if (this.pace !== undefined) payload.pace = this.pace;

    // Per-model controls: pitch / loudness / enable_preprocessing are
    // bulbul:v2-only; temperature / dict_id are bulbul:v3-only. Gate them so we
    // never POST a field the selected model rejects.
    if (this.isV2()) {
      if (this.pitch !== undefined) payload.pitch = this.pitch;
      if (this.loudness !== undefined) payload.loudness = this.loudness;
      if (this.enablePreprocessing !== undefined)
        payload.enable_preprocessing = this.enablePreprocessing;
    } else {
      if (this.temperature !== undefined) payload.temperature = this.temperature;
      if (this.dictId !== undefined) payload.dict_id = this.dictId;
    }

    return payload;
  }

  /**
   * Pre-call HTTP warmup for the Sarvam TTS API.
   *
   * Issues a lightweight `GET` against the API origin so DNS, TLS, and HTTP/2
   * are already up by the time the first `synthesizeStream()` POST lands.
   * Best-effort: 5 s timeout, all exceptions swallowed at debug level.
   *
   * Billing safety: Sarvam does not document a free voice/metadata GET, so this
   * only primes the connection against the API origin root — it never hits the
   * synthesis endpoint, so no characters are billed. Synthesis is billed only
   * when `POST /text-to-speech` runs with non-empty `text`.
   */
  async warmup(): Promise<void> {
    try {
      const origin = new URL(this.baseUrl).origin + '/';
      await fetch(origin, {
        method: 'GET',
        headers: { 'api-subscription-key': this.apiKey },
        signal: AbortSignal.timeout(5_000),
      });
    } catch (err) {
      getLogger().debug(`Sarvam TTS warmup failed (best-effort): ${String(err)}`);
    }
  }

  /** Synthesize text and return the concatenated audio buffer. */
  async synthesize(text: string): Promise<Buffer> {
    const chunks: Buffer[] = [];
    for await (const chunk of this.synthesizeStream(text)) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks);
  }

  /**
   * Synthesize text and yield decoded audio chunks. With the default
   * `codec="linear16"` these are raw PCM_S16LE bytes at `sampleRate`. The REST
   * endpoint returns the whole clip in one JSON response (`audios` is a list of
   * base64 strings); we decode and yield each entry in order.
   */
  async *synthesizeStream(text: string): AsyncGenerator<Buffer> {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'api-subscription-key': this.apiKey,
      },
      body: JSON.stringify(this.buildPayload(text)),
      signal: AbortSignal.timeout(30_000),
    });

    if (!response.ok) {
      const body = await response.text();
      throw new Error(`Sarvam TTS error ${response.status}: ${body}`);
    }

    const data = (await response.json()) as unknown;
    for (const audioB64 of iterAudioB64(data)) {
      const decoded = Buffer.from(audioB64, 'base64');
      if (decoded.length > 0) yield decoded;
    }
  }
}

/**
 * Extract the base64 audio strings from a Sarvam REST response. Accepts the
 * documented `{ audios: [...] }` shape and is defensive about a single-string
 * `audios` value. Returns `[]` for anything unexpected so the caller yields
 * nothing rather than throwing on a malformed body.
 */
function iterAudioB64(data: unknown): string[] {
  if (typeof data !== 'object' || data === null) return [];
  const audios = (data as { audios?: unknown }).audios;
  if (typeof audios === 'string') return audios ? [audios] : [];
  if (Array.isArray(audios)) {
    return audios.filter((a): a is string => typeof a === 'string' && a.length > 0);
  }
  return [];
}
